// @howljs/calendar-kit/lib/module/components/ResourceHeaderItem.js
import React, { useCallback } from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { useLocale } from '../context/LocaleProvider';
import { useTheme } from '../context/ThemeProvider';
import { parseDateTime } from '../utils/dateUtils';
const ResourceHeaderItem = ({
  resources,
  startUnix,
  DateComponent,
  renderResource,
  isShowSeparator = true,
  dateFormat = 'yyyy-MM-dd',
  isShowWeekDay = true
}) => {
  const {
    weekDayShort
  } = useLocale();
  const borderColor = useTheme(useCallback(state => state.colors.border, []));
  const _renderDate = () => {
    const date = parseDateTime(startUnix);
    return /*#__PURE__*/React.createElement(View, {
      style: styles.dateContainer
    }, /*#__PURE__*/React.createElement(Text, {
      style: styles.dateText
    }, `${date.toFormat(dateFormat)}`, isShowWeekDay && ` (${weekDayShort[date.weekday % 7]})`));
  };
  const _renderResource = resource => {
    return /*#__PURE__*/React.createElement(View, {
      key: resource.id,
      style: styles.resource
    }, isShowSeparator && /*#__PURE__*/React.createElement(View, {
      style: {
        width: 1,
        position: 'absolute',
        left: -0.5,
        height: '100%',
        backgroundColor: borderColor
      }
    }), renderResource ? renderResource(resource) : /*#__PURE__*/React.createElement(View, {
      style: [styles.resourceContent, {
        borderColor
      }]
    }, /*#__PURE__*/React.createElement(Text, null, resource.title)));
  };
  return /*#__PURE__*/React.createElement(View, {
    style: styles.container
  }, DateComponent ?? _renderDate(), /*#__PURE__*/React.createElement(View, {
    style: styles.resourceContainer
  }, resources === null || resources === void 0 ? void 0 : resources.map(_renderResource)));
};
export default ResourceHeaderItem;
const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  resourceContainer: {
    flex: 1,
    flexDirection: 'row'
  },
  resource: {
    flex: 1
  },
  resourceContent: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    borderColor: '#ccc',
    borderTopWidth: 1
  },
  dateContainer: {
    paddingVertical: 4
  },
  dateText: {
    textAlign: 'center',
    fontWeight: 'bold'
  }
});
