// @howljs/calendar-kit/lib/commonjs/service/recyclerlistview/core/ScrollComponent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var React = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _reactNativeGestureHandler = require("react-native-gesture-handler");
var _reactNativeReanimated = _interopRequireDefault(require("react-native-reanimated"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const AnimatedScrollView = _reactNativeReanimated.default.createAnimatedComponent(_reactNativeGestureHandler.ScrollView);
class ScrollComponent extends React.Component {
  static defaultProps = {
    contentHeight: 0,
    contentWidth: 0,
    scrollEventThrottle: 16
  };
  _scrollViewRef = null;
  scrollTo(x, y, isAnimated) {
    if (this._scrollViewRef) {
      this._scrollViewRef.scrollTo({
        x,
        y,
        animated: isAnimated
      });
    }
  }
  getScrollableNode() {
    if (this._scrollViewRef && this._scrollViewRef.getScrollableNode) {
      return this._scrollViewRef.getScrollableNode();
    }
    return null;
  }
  render() {
    return /*#__PURE__*/React.createElement(AnimatedScrollView, _extends({
      removeClippedSubviews: false,
      scrollEventThrottle: this.props.scrollEventThrottle
    }, this.props, {
      horizontal: true,
      onScroll: this._onScroll,
      ref: this._getScrollViewRef
    }), /*#__PURE__*/React.createElement(_reactNative.View, {
      style: styles.container
    }, /*#__PURE__*/React.createElement(_reactNative.View, {
      style: {
        height: this.props.contentHeight,
        width: this.props.contentWidth
      }
    }, this.props.children)));
  }
  _getScrollViewRef = scrollView => {
    var _this$props;
    this._scrollViewRef = scrollView;
    (_this$props = this.props) === null || _this$props === void 0 || _this$props.scrollRefExternal(scrollView);
    return scrollView;
  };
  _onScroll = event => {
    if (event) {
      this.props.onScroll(event);
    }
  };
}
exports.default = ScrollComponent;
const styles = _reactNative.StyleSheet.create({
  container: {
    flexDirection: 'row'
  }
});
