// @howljs/calendar-kit/lib/module/service/recyclerlistview/core/ScrollComponent.js
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
import * as React from 'react';
import { StyleSheet, View } from 'react-native';
import { ScrollView } from 'react-native-gesture-handler';
import Animated from 'react-native-reanimated';
const AnimatedScrollView = Animated.createAnimatedComponent(ScrollView);
export default class ScrollComponent extends React.Component {
  static defaultProps = {
    contentHeight: 0,
    contentWidth: 0,
    scrollEventThrottle: 16
  };
  _scrollViewRef = null;
  scrollTo(x, y, isAnimated) {
    if (this._scrollViewRef) {
      this._scrollViewRef.scrollTo({
        x,
        y,
        animated: isAnimated
      });
    }
  }
  getScrollableNode() {
    if (this._scrollViewRef && this._scrollViewRef.getScrollableNode) {
      return this._scrollViewRef.getScrollableNode();
    }
    return null;
  }
  render() {
    return /*#__PURE__*/React.createElement(AnimatedScrollView, _extends({
      removeClippedSubviews: false,
      scrollEventThrottle: this.props.scrollEventThrottle
    }, this.props, {
      horizontal: true,
      onScroll: this._onScroll,
      ref: this._getScrollViewRef
    }), /*#__PURE__*/React.createElement(View, {
      style: styles.container
    }, /*#__PURE__*/React.createElement(View, {
      style: {
        height: this.props.contentHeight,
        width: this.props.contentWidth
      }
    }, this.props.children)));
  }
  _getScrollViewRef = scrollView => {
    var _this$props;
    this._scrollViewRef = scrollView;
    (_this$props = this.props) === null || _this$props === void 0 || _this$props.scrollRefExternal(scrollView);
    return scrollView;
  };
  _onScroll = event => {
    if (event) {
      this.props.onScroll(event);
    }
  };
}
const styles = StyleSheet.create({
  container: {
    flexDirection: 'row'
  }
});
