// @howljs/calendar-kit/lib/commonjs/components/BodyItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));
var _constants = require("../constants");
var _BodyContext = require("../context/BodyContext");
var _DraggableEvent = _interopRequireDefault(require("./DraggableEvent"));
var _Events = _interopRequireDefault(require("./Events"));
var _Overlay = _interopRequireDefault(require("./Loading/Overlay"));
var _NowIndicator = _interopRequireDefault(require("./NowIndicator"));
var _TimelineBoard = _interopRequireDefault(require("./TimelineBoard"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const BodyItem = ({
  pageIndex,
  startUnix,
  renderDraggableEvent,
  resources
}) => {
  const {
    spaceFromTop,
    timelineHeight,
    spaceFromBottom,
    hourWidth,
    numberOfDays,
    calendarData,
    columns
  } = (0, _BodyContext.useBody)();
  const visibleDates = (0, _react.useMemo)(() => {
    const data = {};
    let diffDays = 1;
    for (let i = 0; i < columns; i++) {
      const currentUnix = calendarData.visibleDatesArray[pageIndex + i];
      if (currentUnix) {
        data[currentUnix] = {
          unix: currentUnix,
          diffDays
        };
        diffDays += 1;
      }
    }
    return data;
  }, [calendarData.visibleDatesArray, columns, pageIndex]);
  const leftSpacing = numberOfDays === 1 ? hourWidth : 0;
  const height = (0, _reactNativeReanimated.useDerivedValue)(() => {
    return timelineHeight.value - spaceFromTop - spaceFromBottom;
  }, [spaceFromTop, spaceFromBottom]);
  const animView = (0, _reactNativeReanimated.useAnimatedStyle)(() => ({
    height: height.value
  }));
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.container
  }, /*#__PURE__*/_react.default.createElement(_TimelineBoard.default, {
    pageIndex: pageIndex,
    dateUnix: startUnix,
    visibleDates: visibleDates,
    resources: resources
  }), /*#__PURE__*/_react.default.createElement(_reactNativeReanimated.default.View, {
    pointerEvents: "box-none",
    style: [styles.content, {
      left: resources ? 0 : Math.max(0, leftSpacing - 1),
      top: _constants.EXTRA_HEIGHT + spaceFromTop
    }, animView]
  }, /*#__PURE__*/_react.default.createElement(_Events.default, {
    startUnix: startUnix,
    visibleDates: visibleDates,
    totalResources: resources === null || resources === void 0 ? void 0 : resources.length
  }), /*#__PURE__*/_react.default.createElement(_NowIndicator.default, {
    startUnix: startUnix,
    visibleDates: visibleDates
  }), /*#__PURE__*/_react.default.createElement(_DraggableEvent.default, {
    startUnix: startUnix,
    visibleDates: visibleDates,
    renderDraggableEvent: renderDraggableEvent,
    resources: resources
  })), /*#__PURE__*/_react.default.createElement(_Overlay.default, null));
};
var _default = exports.default = BodyItem;
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1
  },
  content: {
    position: 'absolute',
    width: '100%'
  }
});
