// @howljs/calendar-kit/lib/commonjs/components/Events.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _ActionsProvider = require("../context/ActionsProvider");
var _BodyContext = require("../context/BodyContext");
var _DragEventProvider = require("../context/DragEventProvider");
var _EventsProvider = require("../context/EventsProvider");
var _TimeZoneProvider = require("../context/TimeZoneProvider");
var _dateUtils = require("../utils/dateUtils");
var _EventItem = _interopRequireDefault(require("./EventItem"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const Events = ({
  startUnix,
  visibleDates,
  totalResources
}) => {
  const {
    renderEvent,
    numberOfDays,
    columnWidth
  } = (0, _BodyContext.useBody)();
  const {
    onPressEvent,
    onLongPressEvent
  } = (0, _ActionsProvider.useActions)();
  const {
    draggingId,
    selectedEventId
  } = (0, _DragEventProvider.useDragEvent)();
  const {
    timeZone
  } = (0, _TimeZoneProvider.useTimezone)();
  const {
    triggerDragEvent
  } = (0, _DragEventProvider.useDragEventActions)();
  const {
    data: events
  } = (0, _EventsProvider.useRegularEvents)(startUnix, numberOfDays, visibleDates);
  const _triggerDragEvent = (0, _react.useCallback)((event, resEvent) => {
    if (!event.start.dateTime || !event.end.dateTime) {
      return;
    }
    const eventStart = (0, _dateUtils.forceUpdateZone)(event._internal.startUnix, timeZone).startOf('day');
    const originalStart = (0, _dateUtils.forceUpdateZone)((0, _dateUtils.parseDateTime)(event.start.dateTime, {
      zone: event.start.timeZone
    }), timeZone).startOf('day');
    const startIndex = eventStart.diff(originalStart, 'days').days;
    triggerDragEvent({
      start: {
        dateTime: event.start.dateTime,
        timeZone: event.start.timeZone
      },
      end: {
        dateTime: event.end.dateTime,
        timeZone: event.end.timeZone
      },
      startIndex,
      dragX: event._internal.resourceIndex && totalResources ? event._internal.resourceIndex * (columnWidth / totalResources) + resEvent.nativeEvent.locationX : undefined
    }, event);
  }, [timeZone, triggerDragEvent, columnWidth, totalResources]);
  const _onLongPressEvent = (0, _react.useCallback)(event => {
    const clonedEvent = {
      ...event,
      _internal: undefined
    };
    delete clonedEvent._internal;
    onLongPressEvent === null || onLongPressEvent === void 0 || onLongPressEvent(clonedEvent);
  }, [onLongPressEvent]);
  if (events.length === 0) {
    return null;
  }
  const _renderEvent = event => {
    return /*#__PURE__*/_react.default.createElement(_EventItem.default, {
      key: event.localId,
      event: event,
      startUnix: startUnix,
      renderEvent: renderEvent,
      onPressEvent: onPressEvent,
      onLongPressEvent: triggerDragEvent ? _triggerDragEvent : onLongPressEvent ? _onLongPressEvent : undefined,
      isDragging: draggingId === event.localId || selectedEventId === event.localId,
      visibleDates: visibleDates,
      totalResources: totalResources
    });
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    pointerEvents: "box-none",
    style: _reactNative.StyleSheet.absoluteFill
  }, events.map(_renderEvent));
};
var _default = exports.default = Events;
