// @howljs/calendar-kit/lib/commonjs/components/WeekNumber.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _reactNativeReanimated = require("react-native-reanimated");
var _ThemeProvider = require("../context/ThemeProvider");
var _dateUtils = require("../utils/dateUtils");
var _Text = _interopRequireDefault(require("./Text"));
var _TimeZoneProvider = require("../context/TimeZoneProvider");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const WeekNumber = ({
  date
}) => {
  const {
    timeZone
  } = (0, _TimeZoneProvider.useTimezone)();
  const theme = (0, _ThemeProvider.useTheme)((0, _react.useCallback)(state => ({
    weekNumberBackgroundColor: state.colors.surface,
    weekNumber: state.weekNumber,
    weekNumberContainer: state.weekNumberContainer
  }), []));
  const [value, setValue] = (0, _react.useState)('');
  const _getWeekNumber = newValue => {
    setValue((0, _dateUtils.getWeekNumberOfYear)(newValue, timeZone));
  };
  (0, _reactNativeReanimated.useAnimatedReaction)(() => date.value, (newValue, prevValue) => {
    if (newValue !== prevValue) {
      (0, _reactNativeReanimated.runOnJS)(_getWeekNumber)(newValue);
    }
  }, []);
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, {
      backgroundColor: theme.weekNumberBackgroundColor
    }, theme.weekNumberContainer]
  }, /*#__PURE__*/_react.default.createElement(_Text.default, {
    style: [styles.text, theme.weekNumber]
  }, value));
};
var _default = exports.default = WeekNumber;
const styles = _reactNative.StyleSheet.create({
  container: {
    backgroundColor: '#DADADA',
    marginHorizontal: 8,
    borderRadius: 4,
    marginTop: 8
  },
  text: {
    fontSize: 12,
    textAlign: 'center'
  }
});
