// @howljs/calendar-kit/lib/commonjs/components/TimelineBoard/OutOfRangeView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));
var _BodyContext = require("../../context/BodyContext");
var _ThemeProvider = require("../../context/ThemeProvider");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const OutOfRangeView = ({
  position,
  diffDays
}) => {
  const {
    columnWidthAnim,
    renderCustomOutOfRange,
    timeIntervalHeight,
    totalSlots
  } = (0, _BodyContext.useBody)();
  const disableBackgroundColor = (0, _ThemeProvider.useTheme)(state => state.outOfRangeBackgroundColor || state.colors.surface);
  const disableHeight = (0, _reactNativeReanimated.useDerivedValue)(() => timeIntervalHeight.value * totalSlots);
  const disableWidth = (0, _reactNativeReanimated.useDerivedValue)(() => diffDays * columnWidthAnim.value);
  const disableAnim = (0, _reactNativeReanimated.useAnimatedStyle)(() => {
    return {
      width: disableWidth.value,
      height: disableHeight.value
    };
  });
  const positionStyle = position === 'left' ? {
    left: 0
  } : {
    right: 0
  };
  return /*#__PURE__*/_react.default.createElement(_reactNativeReanimated.default.View, {
    style: [styles.outOfRange, positionStyle, {
      backgroundColor: disableBackgroundColor
    }, disableAnim]
  }, renderCustomOutOfRange && renderCustomOutOfRange({
    width: disableWidth,
    height: disableHeight
  }));
};
var _default = exports.default = OutOfRangeView;
const styles = _reactNative.StyleSheet.create({
  outOfRange: {
    position: 'absolute',
    height: '100%'
  }
});
