// @howljs/calendar-kit/lib/commonjs/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScrollTypeMonth = exports.ScrollType = exports.MIN_DATE = exports.MIN_ALL_DAY_MINUTES = exports.MINUTES_IN_DAY = exports.MILLISECONDS_IN_MINUTE = exports.MILLISECONDS_IN_DAY = exports.MAX_DATE = exports.MAX_ALL_DAY_MINUTES = exports.INITIAL_DATE = exports.HOUR_WIDTH = exports.HOUR_SHORT_LINE_WIDTH = exports.HEADER_BOTTOM_HEIGHT = exports.EXTRA_HEIGHT = exports.DEFAULT_THEME = exports.DEFAULT_SIZE = exports.DEFAULT_MIN_START_DIFFERENCE = exports.DEFAULT_LOCALES = exports.DEFAULT_ALL_DAY_MINUTES = exports.DEBOUNCE_TIME = exports.DAY_BAR_HEIGHT = exports.COLLAPSED_ITEMS = void 0;
const DEFAULT_THEME = exports.DEFAULT_THEME = {
  colors: {
    primary: '#1a73e8',
    onPrimary: '#fff',
    background: '#fff',
    onBackground: '#000',
    border: '#dadce0',
    text: '#000',
    surface: '#ECECEC',
    onSurface: '#000'
  }
};
const HOUR_WIDTH = exports.HOUR_WIDTH = 60;
const DAY_BAR_HEIGHT = exports.DAY_BAR_HEIGHT = 60;
const INITIAL_DATE = exports.INITIAL_DATE = new Date().toISOString();
const MIN_DATE = exports.MIN_DATE = new Date(new Date().getFullYear() - 2, new Date().getMonth(), new Date().getDate()).toISOString();
const MAX_DATE = exports.MAX_DATE = new Date(new Date().getFullYear() + 2, new Date().getMonth(), new Date().getDate()).toISOString();
const DEFAULT_LOCALES = exports.DEFAULT_LOCALES = {
  en: {
    weekDayShort: 'Sun_Mon_Tue_Wed_Thu_Fri_Sat'.split('_'),
    meridiem: {
      ante: 'am',
      post: 'pm'
    },
    more: '{count} more'
  }
};
const MILLISECONDS_IN_DAY = exports.MILLISECONDS_IN_DAY = 86400000;
const MILLISECONDS_IN_MINUTE = exports.MILLISECONDS_IN_MINUTE = 60000;
const MINUTES_IN_DAY = exports.MINUTES_IN_DAY = 1440;
let ScrollType = exports.ScrollType = /*#__PURE__*/function (ScrollType) {
  ScrollType[ScrollType["dayBar"] = 0] = "dayBar";
  ScrollType[ScrollType["calendarGrid"] = 1] = "calendarGrid";
  return ScrollType;
}({});
let ScrollTypeMonth = exports.ScrollTypeMonth = /*#__PURE__*/function (ScrollTypeMonth) {
  ScrollTypeMonth[ScrollTypeMonth["header"] = 0] = "header";
  ScrollTypeMonth[ScrollTypeMonth["body"] = 1] = "body";
  return ScrollTypeMonth;
}({});
const EXTRA_HEIGHT = exports.EXTRA_HEIGHT = 300;
const HOUR_SHORT_LINE_WIDTH = exports.HOUR_SHORT_LINE_WIDTH = 8;
const COLLAPSED_ITEMS = exports.COLLAPSED_ITEMS = 2;
const HEADER_BOTTOM_HEIGHT = exports.HEADER_BOTTOM_HEIGHT = 20;
const DEFAULT_SIZE = exports.DEFAULT_SIZE = {
  width: 0,
  height: 0
};
const DEBOUNCE_TIME = exports.DEBOUNCE_TIME = 200;
const MIN_ALL_DAY_MINUTES = exports.MIN_ALL_DAY_MINUTES = 20;
const MAX_ALL_DAY_MINUTES = exports.MAX_ALL_DAY_MINUTES = 30;
const DEFAULT_ALL_DAY_MINUTES = exports.DEFAULT_ALL_DAY_MINUTES = 20;
const DEFAULT_MIN_START_DIFFERENCE = exports.DEFAULT_MIN_START_DIFFERENCE = 30;
