// @howljs/calendar-kit/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  CalendarBody: true,
  CalendarContainer: true,
  CalendarHeader: true,
  ResourceHeaderItem: true,
  MultiDayBarItem: true,
  SingleDayBarItem: true,
  DayItem: true,
  useActions: true,
  useBody: true,
  useCalendar: true,
  useDayBar: true,
  useHeader: true,
  useDragEvent: true,
  useDragEventActions: true,
  useAllDayEvents: true,
  useAllDayEventsByDay: true,
  useEventCountsByWeek: true,
  useMonthEvents: true,
  useRegularEvents: true,
  useHighlightDates: true,
  useLayout: true,
  useLoading: true,
  useLocale: true,
  useNowIndicator: true,
  useTheme: true,
  useTimezone: true,
  useUnavailableHours: true,
  useUnavailableHoursByDate: true,
  useDateChangedListener: true,
  useNotifyDateChanged: true,
  DraggableEvent: true,
  DraggingEvent: true
};
Object.defineProperty(exports, "CalendarBody", {
  enumerable: true,
  get: function () {
    return _CalendarBody.default;
  }
});
Object.defineProperty(exports, "CalendarContainer", {
  enumerable: true,
  get: function () {
    return _CalendarContainer.default;
  }
});
Object.defineProperty(exports, "CalendarHeader", {
  enumerable: true,
  get: function () {
    return _CalendarHeader.default;
  }
});
Object.defineProperty(exports, "DayItem", {
  enumerable: true,
  get: function () {
    return _DayItem.default;
  }
});
Object.defineProperty(exports, "DraggableEvent", {
  enumerable: true,
  get: function () {
    return _DraggableEvent.DraggableEvent;
  }
});
Object.defineProperty(exports, "DraggingEvent", {
  enumerable: true,
  get: function () {
    return _DraggingEvent.DraggingEvent;
  }
});
Object.defineProperty(exports, "MultiDayBarItem", {
  enumerable: true,
  get: function () {
    return _MultiDayBarItem.default;
  }
});
Object.defineProperty(exports, "ResourceHeaderItem", {
  enumerable: true,
  get: function () {
    return _ResourceHeaderItem.default;
  }
});
Object.defineProperty(exports, "SingleDayBarItem", {
  enumerable: true,
  get: function () {
    return _SingleDayBarItem.default;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "useActions", {
  enumerable: true,
  get: function () {
    return _ActionsProvider.useActions;
  }
});
Object.defineProperty(exports, "useAllDayEvents", {
  enumerable: true,
  get: function () {
    return _EventsProvider.useAllDayEvents;
  }
});
Object.defineProperty(exports, "useAllDayEventsByDay", {
  enumerable: true,
  get: function () {
    return _EventsProvider.useAllDayEventsByDay;
  }
});
Object.defineProperty(exports, "useBody", {
  enumerable: true,
  get: function () {
    return _BodyContext.useBody;
  }
});
Object.defineProperty(exports, "useCalendar", {
  enumerable: true,
  get: function () {
    return _CalendarProvider.useCalendar;
  }
});
Object.defineProperty(exports, "useDateChangedListener", {
  enumerable: true,
  get: function () {
    return _VisibleDateProvider.useDateChangedListener;
  }
});
Object.defineProperty(exports, "useDayBar", {
  enumerable: true,
  get: function () {
    return _DayBarContext.useHeader;
  }
});
Object.defineProperty(exports, "useDragEvent", {
  enumerable: true,
  get: function () {
    return _DragEventProvider.useDragEvent;
  }
});
Object.defineProperty(exports, "useDragEventActions", {
  enumerable: true,
  get: function () {
    return _DragEventProvider.useDragEventActions;
  }
});
Object.defineProperty(exports, "useEventCountsByWeek", {
  enumerable: true,
  get: function () {
    return _EventsProvider.useEventCountsByWeek;
  }
});
Object.defineProperty(exports, "useHeader", {
  enumerable: true,
  get: function () {
    return _DayBarContext.useHeader;
  }
});
Object.defineProperty(exports, "useHighlightDates", {
  enumerable: true,
  get: function () {
    return _HighlightDatesProvider.useHighlightDates;
  }
});
Object.defineProperty(exports, "useLayout", {
  enumerable: true,
  get: function () {
    return _LayoutProvider.useLayout;
  }
});
Object.defineProperty(exports, "useLoading", {
  enumerable: true,
  get: function () {
    return _LoadingContext.useLoading;
  }
});
Object.defineProperty(exports, "useLocale", {
  enumerable: true,
  get: function () {
    return _LocaleProvider.useLocale;
  }
});
Object.defineProperty(exports, "useMonthEvents", {
  enumerable: true,
  get: function () {
    return _EventsProvider.useMonthEvents;
  }
});
Object.defineProperty(exports, "useNotifyDateChanged", {
  enumerable: true,
  get: function () {
    return _VisibleDateProvider.useNotifyDateChanged;
  }
});
Object.defineProperty(exports, "useNowIndicator", {
  enumerable: true,
  get: function () {
    return _NowIndicatorProvider.useNowIndicator;
  }
});
Object.defineProperty(exports, "useRegularEvents", {
  enumerable: true,
  get: function () {
    return _EventsProvider.useRegularEvents;
  }
});
Object.defineProperty(exports, "useTheme", {
  enumerable: true,
  get: function () {
    return _ThemeProvider.useTheme;
  }
});
Object.defineProperty(exports, "useTimezone", {
  enumerable: true,
  get: function () {
    return _TimeZoneProvider.useTimezone;
  }
});
Object.defineProperty(exports, "useUnavailableHours", {
  enumerable: true,
  get: function () {
    return _UnavailableHoursProvider.useUnavailableHours;
  }
});
Object.defineProperty(exports, "useUnavailableHoursByDate", {
  enumerable: true,
  get: function () {
    return _UnavailableHoursProvider.useUnavailableHoursByDate;
  }
});
var _CalendarKit = _interopRequireDefault(require("./CalendarKit"));
var _CalendarBody = _interopRequireDefault(require("./CalendarBody"));
var _CalendarContainer = _interopRequireDefault(require("./CalendarContainer"));
var _CalendarHeader = _interopRequireDefault(require("./CalendarHeader"));
var _ResourceHeaderItem = _interopRequireDefault(require("./components/ResourceHeaderItem"));
var _MultiDayBarItem = _interopRequireDefault(require("./components/MultiDayBarItem"));
var _SingleDayBarItem = _interopRequireDefault(require("./components/SingleDayBarItem"));
var _DayItem = _interopRequireDefault(require("./components/DayItem"));
var _ActionsProvider = require("./context/ActionsProvider");
var _BodyContext = require("./context/BodyContext");
var _CalendarProvider = require("./context/CalendarProvider");
var _DayBarContext = require("./context/DayBarContext");
var _DragEventProvider = require("./context/DragEventProvider");
var _EventsProvider = require("./context/EventsProvider");
var _HighlightDatesProvider = require("./context/HighlightDatesProvider");
var _LayoutProvider = require("./context/LayoutProvider");
var _LoadingContext = require("./context/LoadingContext");
var _LocaleProvider = require("./context/LocaleProvider");
var _NowIndicatorProvider = require("./context/NowIndicatorProvider");
var _ThemeProvider = require("./context/ThemeProvider");
var _TimeZoneProvider = require("./context/TimeZoneProvider");
var _UnavailableHoursProvider = require("./context/UnavailableHoursProvider");
var _VisibleDateProvider = require("./context/VisibleDateProvider");
var _DraggableEvent = require("./components/DraggableEvent");
var _DraggingEvent = require("./components/DraggingEvent");
var _dateUtils = require("./utils/dateUtils");
Object.keys(_dateUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _dateUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _dateUtils[key];
    }
  });
});
var _rrule = require("./service/rrule");
Object.keys(_rrule).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _rrule[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _rrule[key];
    }
  });
});
var _types = require("./types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
var _default = exports.default = _CalendarKit.default; // Recommend: use useHeader instead. useDayBar will be removed soon.
